### PR TITLE
CDAP-13088 Metadata from program and plugin docs [Part 2]

### PR DESCRIPTION
**Pull request by rohitsinha54**
### Part 2

- Documents accessing Metadata from Plugin and Pipelines.

Build: https://builds.cask.co/browse/CDAP-DQB412-1

Rendered Page: https://builds.cask.co/artifact/CDAP-DQB412/shared/build-4/Docs-HTML/5.0.0-SNAPSHOT/en/developer-manual/metadata/programmatic-metadata.html

### CODE REVIEW

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+.. _programmatic-metadata:
+
+==================================
+Metadata from Programs and Plugins
+==================================
+
+.. highlight:: java
+
+Metadata can be accessed from *Programs* and *Plugins*. Currently, reading metadata is not supported while CDAP is running in *cloud mode* but metadata can be added, updated or deleted in all modes.
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: Can you add a newline after 120 characters? Makes it easier to review. 

Also, I don't think CDAP has a cloud mode. Perhaps change it to ...reading metadata from a program or a plugin is not supported when the program is running in the cloud....

Although, is that a true statement? Is it supported when you run the program on a remote Hadoop cluster, for example?

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+.. _metadata-programs:
+
+Programs
+========
+
+Metadata can be accessed from *MapReduce*, *Spark*, *Workers* and *Service* through their individual program context.
+Metadata operations are not supported at individual tasks and should be performed only during the
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: is this more a best practice? Or does the API not allow this to happen?

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+initialization phase of the program.
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: good to indicate the method name as well

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+    // construct the metadata entity
+    MetadataEntity entity = MetadataEntity.builder().append(MetadataEntity.NAMESPACE, "myNamespace")
+      .appendAsType(MetadataEntity.DATASET, "myDataset").build();
+
+    // get the metadata
+    Map<MetadataScope, Metadata> metadata = context.getMetadata(entity);
+  }
+
+A program can also emit metadata properties and tags using the context.
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: A program can also *annotate metadata to an entity using the same context object*

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+    // add a tag
+    context.addTags(entity, "someTag");
+  }
+
+.. _metadata-plugins:
+
+Plugins
+=======
+A *Plugin* can also access metadata through the context provided to it in ``prepareRun`` stage. Like *programs*, metadata operations are not supported at individual task level.
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: break at 120 characters.

Comment (bdmogal) [reply cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: Same question as above as well.

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+    // construct the metadata entity
+    MetadataEntity entity = MetadataEntity.builder().append(MetadataEntity.NAMESPACE, "myNamespace")
+      .appendAsType(MetadataEntity.DATASET, "myDataset").build();
+
+    // get the metadata
+    Map<MetadataScope, Metadata> metadata = context.getMetadata(entity);
+  }
+
+A plugin can also emit metadata. For example a plugin can tag a dataset in the following way::
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: similar change about annotating metadata to an entity here

**Review by bdmogal — CHANGES_REQUESTED** [review]
@rohitsinha54 for both these cases, it would be good to add a reference to the javadocs.

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst (hunk 1 above)
Comment (rohitsinha54) [reply cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: You can read metadata while running on hadoop cluster (in-prem). We use service discovery to discover metadata service and retrieve metadata through that. 

By remote hadoop cluster you mean the newly added hadoop provisioner?

Comment (rohitsinha54) [reply cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: Done first 2 changes mentioned in this comment.

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst (hunk 2 above)
Comment (rohitsinha54) [reply cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: We don't even expose these api to task level so I will just remove this line.

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst (hunk 3 above)
Comment (rohitsinha54) [reply cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: Done.

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst (hunk 5 above)
Comment (rohitsinha54) [reply cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: Done.

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst (hunk 4 above)
Comment (rohitsinha54) [reply cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: Done.

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst (hunk 6 above)
Comment (rohitsinha54) [reply cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: Done.

**Comment by rohitsinha54** [comment]
@bdmogal Thanks for the review. I have addressed the comments. Please have another look.

File: cdap-docs/developer-manual/source/metadata/index.rst [hunk 7]
```diff
@@ -38,7 +39,7 @@ and governance for every application.
 
 CDAP metadata |---| consisting of **properties** (a list of key-value pairs) or **tags** (a
 list of keys) |---| can be used to annotate artifacts, applications, programs, datasets,
-streams, and views.
+streams, views and many other entities including custom entities.
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/metadata/index.rst]: ... and custom entities.

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+.. _programmatic-metadata:
+
+==================================
+Metadata from Programs and Plugins
+==================================
+
+.. highlight:: java
+
+Metadata can be accessed from *Programs* and *Plugins*. Currently, reading metadata from a program or plugin is not
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: I think this page is missing a very important point - why would someone do this? Shall we cite one or two example use cases?

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+supported while it is running in *cloud* but metadata can be added, updated or deleted.
+
+.. _metadata-programs:
+
+Programs
+========
+
+Metadata can be accessed from *MapReduce*, *Spark*, *Workers* and *Service* through
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: ... through methods from the MetadataReader object, which are available via the appropriate program context object in your program. The program context object can be obtained by invoking the ``getContext()`` method inside the ``initialize`` method of your program.

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+`MetadataReader <../../reference-manual/javadocs/co/cask/cdap/api/metadata/MetadataReader.html>`__ exposed to
+individual program context available in the ``initialize`` method.
+
+For example metadata can be read in *MapReduce* programs in the following way::
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: The following example shows how you can retrieve the metadata of an entity in a *MapReduce* program::

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+    // construct the metadata entity
+    MetadataEntity entity = MetadataEntity.builder().append(MetadataEntity.NAMESPACE, "myNamespace")
+      .appendAsType(MetadataEntity.DATASET, "myDataset").build();
+
+    // get the metadata
+    Map<MetadataScope, Metadata> metadata = context.getMetadata(entity);
+  }
+
+A program can also annotate metadata to an entity using the
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: You can also annotate metadata to an entity through methods from the MetadataWriter... object, which are also available through the same program context object as above.

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+`MetadataWriter <../../reference-manual/javadocs/co/cask/cdap/api/metadata/MetadataWriter.html>`__
+exposed to same context object. For example a dataset can be tagged from a *MapReduce* program in the following way::
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: for consistency with above, let's add a line break after the period?

Comment (bdmogal) [reply cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: also change the sentence similar to the previous context?

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+    // add a tag
+    context.addTags(entity, "someTag");
+  }
+
+.. _metadata-plugins:
+
+Plugins
+=======
+A *Plugin* can also access metadata through
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: can you also make similar changes here?

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+
+    // construct the metadata entity
+    MetadataEntity entity = MetadataEntity.builder().append(MetadataEntity.NAMESPACE, "myNamespace")
+      .appendAsType(MetadataEntity.DATASET, "myDataset").build();
+
+    // get the metadata
+    Map<MetadataScope, Metadata> metadata = context.getMetadata(entity);
+  }
+
+A plugin can also annotate metadata to an entity using the
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: and here...

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+`MetadataWriter <../../reference-manual/javadocs/co/cask/cdap/api/metadata/MetadataWriter.html>`__
+exposed to same context object. For example a plugin can tag a dataset in the following way::
```
Comment (bdmogal) [inline cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: line break

**Review by bdmogal — CHANGES_REQUESTED** [review]
@rohitsinha54 apologies, I didn't quite understand the API earlier, so added more comments now. Basically, a user never sees the MetadataReader and MetadataWriter objects right? I saw the emphasis on those classes, and then tried to find them in the example code, but realized that a user never sees them. So I made some more suggestions to reflect that. Let me know what you think.

File: cdap-docs/developer-manual/source/metadata/index.rst (hunk 7 above)
Comment (rohitsinha54) [reply cdap-docs/developer-manual/source/metadata/index.rst]: Done.

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst (hunk 10 above)
Comment (rohitsinha54) [reply cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: Done.

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst (hunk 12 above)
Comment (rohitsinha54) [reply cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: Done.

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst (hunk 9 above)
Comment (rohitsinha54) [reply cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: Done.

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst (hunk 11 above)
Comment (rohitsinha54) [reply cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: Done.

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst (hunk 13 above)
Comment (rohitsinha54) [reply cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: Done.

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst (hunk 14 above)
Comment (rohitsinha54) [reply cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: Done.

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst (hunk 15 above)
Comment (rohitsinha54) [reply cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: Done.

File: cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst (hunk 8 above)
Comment (rohitsinha54) [reply cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst]: Done.

**Comment by bdmogal** [comment]
@rohitsinha54  pushed a commit because I wanted to try out a few options. Also fixed some broken links. Please review and squash/merge if it looks good.

**Comment by rohitsinha54** [comment]
@bdmogal Thanks for the review and commit. I am trying to see which link you fixed but can't seem to find it. Can you point to the link?

**Pull request merged**